### PR TITLE
#patch create gcp log formatter

### PR DESCRIPTION
**Pull request by ckiosidis**
Signed-off-by: Babis Kiosidis <babisk@spotify.com>

# TL;DR

Created a structured log that is compatible with GCP logging and tried it from a local Flyteadmin.

I will also try the changes on a k8s pod running on GCP. GCP adds a lot of labels and metadata on the logs and it would be good to verify that it works as expected in that environment

solved https://github.com/flyteorg/flyte/issues/1852

## Type
 - [x] Bug Fix
 - [ ] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [ ] Smoke tested
 - [ ] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
 _How did you fix the bug, make the feature etc. Link to any design docs etc_
Introduced a new formatter of type GCP that maps the logrus.LEVEL to GcpSeverity

## Tracking Issue
_Remove the '*fixes*' keyword if there will be multiple PRs to fix the linked issue_

fixes https://github.com/flyteorg/flyte/issues/<number>

## Follow-up issue
_NA_
OR
_https://github.com/flyteorg/flyte/issues/<number>_

gcp log explorer without formatting
Screenshot 2021-11-22 at 16 54 36

gcp log explorer after formatting

Screenshot 2021-11-23 at 16 10 58
Screenshot 2021-11-23 at 16 11 12

pod logs:

```

kubectl logs flyteadmin flyteadmin
time="2021-11-23T13:58:44Z" level=info msg="Using config file: [/etc/flyteadmin_config.yaml]"
{"data":{"src":"viper.go:398"},"message":"Config section [database] updated. No update handler registered.","severity":"DEBUG","timestamp":"2021-11-23T13:58:44Z"}
{"data":{"src":"viper.go:398"},"message":"Config section [auth] updated. No update handler registered.","severity":"DEBUG","timestamp":"2021-11-23T13:58:44Z"}
{"data":{"src":"viper.go:398"},"message":"Config section [secrets] updated. No update handler registered.","severity":"DEBUG","timestamp":"2021-11-23T13:58:44Z"}
{"data":{"src":"viper.go:398"},"message":"Config section [clusters] updated. No update handler registered.","severity":"DEBUG","timestamp":"2021-11-23T13:58:44Z"}
{"data":{"src":"viper.go:398"},"message":"Config section [queues] updated. No update handler registered.","severity":"DEBUG","timestamp":"2021-11-23T13:58:44Z"}
{"data":{"src":"viper.go:398"},"message":"Config section [qualityofservice] updated. No update handler registered.","severity":"DEBUG","timestamp":"2021-11-23T13:58:44Z"}
{"data":{"src":"viper.go:398"},"message":"Config section [flyteadmin] updated. No update handler registered.","severity":"DEBUG","timestamp":"2021-11-23T13:58:44Z"}
{"data":{"src":"viper.go:398"},"message":"Config section [remotedata] updated. No update handler registered.","severity":"DEBUG","timestamp":"2021-11-23T13:58:44Z"}
{"data":{"src":"serve.go:210"},"message":"Serving Flyte Admin Insecure","severity":"INFO","timestamp":"2021-11-23T13:58:44Z"}
{"data":{"src":"serve.go:108"},"message":"Creating gRPC server without authentication","severity":"INFO","timestamp":"2021-11-23T13:58:44Z"}
{"data":{"src":"application_config_provider.go:103"},"message":"missing database password at specified path [/etc/db-user-pass.conf]","severity":"CRITICAL","timestamp":"2021-11-23T13:58:44Z"}


```

### CODE REVIEW

**Comment by narape** [comment]
@ckiosidis Unit tests are missing, I guess you will add it after testing it locally, right?

File: logger/logger.go [hunk 1]
```diff
@@ -47,6 +52,55 @@ func onConfigUpdated(cfg Config) {
...
 
+type GcpFormatter struct {
+}
+
+type GcpEntry struct {
+	Data      map[string]interface{} `json:"data,omitempty"`
+	Message   string                 `json:"message,omitempty"`
+	Severity  GcpSeverity            `json:"severity,omitempty"`
+	Timestamp string                 `json:"timestamp,omitempty"`
+	// CHECK: Context
```
Comment (narape) [inline logger/logger.go]: It is not clear what is happening here. Is this a TODO that would be addressed in an followup commit or in a follow PR?

**Review by narape — DISMISSED** [review]

**Comment by kumare3** [comment]
Cc @jeevb can you pr too? Or @pradithya

File: logger/logger.go [hunk 2]
```diff
@@ -47,6 +52,54 @@ func onConfigUpdated(cfg Config) {
...
+		logrus.FatalLevel: GcpSeverityCritical,
+		logrus.PanicLevel: GcpSeverityAlert,
+	}
+)
+
+func (f *GcpFormatter) Format(entry *logrus.Entry) ([]byte, error) {
+	var log = &GcpEntry{
+		Timestamp: entry.Time.Format(time.RFC3339),
+		Message:   entry.Message,
+		Severity:  logrusToGcp[entry.Level],
```
Comment (EngHabu) [inline logger/logger.go:92]: You might want to make this default to Debug or something in the rare case we add a log level but not add a mapping here...

File: logger/logger.go [hunk 3]
```diff
@@ -47,6 +52,54 @@ func onConfigUpdated(cfg Config) {
 	}
 }
 
+type GcpFormatter struct {
```
Comment (EngHabu) [inline logger/logger.go:55]: Mind moving the Gcp types to a new file?

**Review by EngHabu — COMMENTED** [review]
Looks good to me. a couple of nits

File: logger/logger.go (hunk 2 above)
Comment (ckiosidis) [reply logger/logger.go]: makes sense

**Comment by ckiosidis** [comment]
fixed your comments @EngHabu PTAL

File: logger/log_formatter.go [hunk 4]
```diff
@@ -0,0 +1,65 @@
+package logger
+
```
Comment (EngHabu) [inline logger/log_formatter.go:2]: Apologies for the nits :-) mind naming this file `gcp_formatter.go` ?

File: logger/log_formatter.go [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+		logrus.DebugLevel: GcpSeverityDebug,
+		logrus.InfoLevel:  GcpSeverityInfo,
+		logrus.WarnLevel:  GcpSeverityWarning,
+		logrus.ErrorLevel: GcpSeverityError,
+		logrus.FatalLevel: GcpSeverityCritical,
+		logrus.PanicLevel: GcpSeverityAlert,
+	}
+)
+
+func getOrDefault(level logrus.Level) GcpSeverity {
```
Comment (EngHabu) [inline logger/log_formatter.go:44]:
```suggestion
func toGcpSeverityOrDefault(level logrus.Level) GcpSeverity {
```

File: logger/log_formatter.go [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+
+import (
+	"encoding/json"
+	"fmt"
+	"time"
+
+	"github.com/sirupsen/logrus"
+)
+
+type GcpFormatter struct {
```
Comment (EngHabu) [inline logger/log_formatter.go:11]: Let's add docs on public types too plz

**Review by EngHabu — COMMENTED** [review]
a couple more nits :-) I hope you don't hate me

File: logger/log_formatter.go (hunk 4 above)
Comment (ckiosidis) [reply logger/log_formatter.go:2]: no problem!

**Review by EngHabu — APPROVED** [review]
Thank you!

**Comment by ckiosidis** [comment]
Thank you! :)
